src/components/Select/CreatableSelect/hooks.js

```diff
@@ -0,0 +1,223 @@
+import { useState, useRef, useCallback, useEffect } from "react";
+import { useClickOutside, useHandleKeyboardHook, usePrevious } from "../hooks";
+import { blankValue } from "../helpers";
+
+export function useCreatableSelect(
+  disabled,
+  autoFocus,
+  onChange,
+  options,
+  defaultValue,
+  getOptionLabel,
+  getOptionValue,
+  filterOptions,
+
+  typeLabel,
+  isValidNewOption,
+  newOptionCreator,
+  promptTextCreator
+) {
+  const optionContainerRef = useRef(null);
+  const dropdownRef = useRef(null);
+
+  const newOptions = useRef([]);
+
+  const [menuOpened, setMenuOpened] = useState(false);
+  const [focusedOptionIndex, setFocusedOptionIndex] = useState(0);
+
+  const defaultValueKeyGetter = useCallback(({ value }) => value, []);
+  const defaultLabelKeyGetter = useCallback(({ label }) => label, []);
+
+  const getValueKey = getOptionValue || defaultValueKeyGetter;
+  const getLabelKey = getOptionLabel || defaultLabelKeyGetter;
+
+  let _options = [...options, ...newOptions.current].filter(
+    (item, pos, self) =>
+      self.findIndex((_item) => getLabelKey(_item) === getLabelKey(item)) ===
+      pos
+  );
+  _options = filterOptions ? _options.filter(filterOptions) : _options;
+  _options.unshift({
+    label: typeLabel || "Type to add a new option",
+    value: ""
+  });
```

Comment:
I think this could be wrapped into a function, or be used in a more declarative approach. Since you are working with fp, it would be better if you use function composition instead of the imperative way.

Reply:
Also, the default value for the `typeLabel` should be declared as a default component prop.